File: quadzone-api/src/main/java/com/quadzone/product/Product.java
package com.quadzone.product;

import com.quadzone.order.OrderItem;
import com.quadzone.product.category.sub_category.SubCategory;
import com.quadzone.product.dto.ProductUpdateRequest;
import com.quadzone.review.Review;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "product")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column
    private String brand;

    @Column(name = "model_number")
    private String modelNumber;

    @Column
    private String color;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private Double price;

    @Column(name = "cost_price")
    private Double costPrice;

    @Column
    private Double weight;

    @Column(name = "stock_quantity", nullable = false)
    private Integer stock;

    @Column(name = "image_url")
    private String imageUrl;

    @Column(name = "created_at", updatable = false)
    private final LocalDateTime createdAt = LocalDateTime.now();

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "is_active")
    private boolean isActive;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "subcategory_id", nullable = false)
    private SubCategory subCategory;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Review> reviews = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<OrderItem> orderItems = new ArrayList<>();

    public void updateFrom(ProductUpdateRequest request) {
        if (request.name() != null) {
            this.setName(request.name());
        }
        if (request.brand() != null) {
            this.setBrand(request.brand());
        }
        if (request.modelNumber() != null) {
            this.setModelNumber(request.modelNumber());
        }
        if (request.description() != null) {
            this.setDescription(request.description());
        }
        if (request.quantity() != null) {
            this.setStock(request.quantity());
        }
        if (request.price() != null) {
            this.setPrice(request.price());
        }
        if (request.costPrice() != null) {
            this.setCostPrice(request.costPrice());
        }
        if (request.weight() != null) {
            this.setWeight(request.weight());
        }
        if (request.color() != null) {
            this.setColor(request.color());
        }
        if (request.imageUrl() != null) {
            this.setImageUrl(request.imageUrl());
        }
        if (request.subCategory() != null) {
            this.setSubCategory(request.subCategory());
        }
    }

    @PreUpdate
    public void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: quadzone-ui/src/types/Product.ts
// Basic product in list/search (ProductResponse from backend)
export interface Product {
    id: number;
    name: string;
    brand: string;
    price: number;
    imageUrl: string;
    createdAt:Date;
    subCategory: SubCategory;
    category: Category;
    isActive:boolean;
}

// Detailed product (ProductDetailsResponse from backend)
export interface ProductDetails extends Product {
    modelNumber: string;
    description: string[];
    weight: number;
    quantity: number;
    reviews: Review[];
}

// Category from backend (CategoryResponse)
export interface Category {
    id: number;
    name: string;
    subCategories: SubCategory[];
}

// SubCategory from backend (SubCategoryResponse)
export interface SubCategory {
    id: number;
    name: string;
}

// Brand from backend (BrandResponse)
export interface Brand {
    brand: string;
}

export interface Review {
    id: number;
    rating: number;
    text: string;
    title?: string;
    userId: number;
    userName?: string;
    createdAt?: string;
}

export interface PriceRange {
    min: number;
    max: number;
}
